### ISSUE TITLE
Restrict question types

### ISSUE THREAD

**Issue by tmtsc** [frankkoch/moodle-mod_studentquiz#393]
First, a question regarding security concerns: The StudentQuiz Plugin integrates various question type plugins that are installed on the Moodle system. It also enables users with student role to access user interfaces which are normally only accessible to teacher roles. How can be guaranteed that every question type plugin works securely when used through the StudentQuiz Plugin (also non-core plugins)?

In this context a restriction setting would be useful, which allows the administrator to restrict the question type plugins that are activated for use with StudentQuiz Plugin. For example, only core plugins could be enabled, as they may be more thoroughly tested interacting with the StudentQuiz Plugin (than external plugins). Also for performance reasons, the administrator could have the requirement to restrict users from using certain question types, e.g. STACK.
Currently, only teachers can set these restrictions in their courses, but not the administrators globally.

**Comment by oucnelson** [comment]
Hi tmtsc,

Assuming I've understood this correctly, I believe we already have this. Under https://[[URL]]/admin/settings.php?section=modsettingstudentquiz, there is the "Allowed question types, studentquiz | defaultqtypes" setting, which sets the question list at the server/website-level. 

When a teacher/tutor/activity designer creates a StudentQuiz instance on a course/module, they can then select/deselect from the specificed qtype list.

Therefore it is already possible to disallow STACK, CodeRunner etc. server/website-level. We've had this feature for quite a while now, and developed the general before the specific setting.

Best wishes,

Chris.

**Comment by timhunt** [comment]
Chris, that setting is only a default, which teachers can change.

The suggestion is to do the same thing for quesition types which was already done for roles, where we have both 'Default roles ...' and 'Exclude roles ...', so that admins can limit what teachers can change.

I think that is a good idea, when we have time to implement it.

**Comment by oucnelson** [comment]
Right, I see the issue now. Apologies - "slow brain" morning (new parent/sleep deprivation ;-) ).

Originally, the server/website setting called "Allowed question types" and should have done as described: limited what qtypes were available for the teacher/tutor to enable at the instance level. 

What we've got under that is studentquiz | defaultqtypes, which sets the defaults qtypes available for all instances that the teacher/tutor/activity designer can then tweak.

So I completely agree we're missing that setting, and we'll try to look at adding that following completion of the M4.0 compatibility work.
